
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Prepare your own data &#8212; ScisTree2 Book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css?v=6644e6bb" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'tutorials/input';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Map mutations on branches" href="branch_length.html" />
    <link rel="prev" title="Basic usage" href="usage.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="ScisTree2 Book - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="ScisTree2 Book - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../index.html">
                    About
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../citation.html">Citation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../installation/installation-py.html">Install for Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation/installation-c.html">Install for C++ (optional)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Basic usage</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Prepare your own data</a></li>
<li class="toctree-l1"><a class="reference internal" href="branch_length.html">Map mutations on branches</a></li>
<li class="toctree-l1"><a class="reference internal" href="bootstrap.html">Get branch support using bootstrap</a></li>
<li class="toctree-l1"><a class="reference internal" href="evaluation.html">Evaluate the inferred tree</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">CLI</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cli.html">Using ScisTree2 in C++</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Release Notes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../release.html">Release Notes</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/yufengwudcs/ScisTree2" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/tutorials/input.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Prepare your own data</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method-1-probability-matrix">Method 1: Probability Matrix</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method-2-from-csv-file">Method 2: From CSV File</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method-3-from-read-counts">Method 3: From Read Counts</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method-4-from-vcf-file">Method 4: From VCF File</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="prepare-your-own-data">
<h1>Prepare your own data<a class="headerlink" href="#prepare-your-own-data" title="Link to this heading">#</a></h1>
<p>ScisTree2 works with genotype probability data, which it accepts as a <code class="docutils literal notranslate"><span class="pre">GenotypeProbability</span></code> object. You have several convenient methods to construct this essential object, each catering to different input data formats.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All data and files can be found in GitHub repo</p>
</div>
<section id="method-1-probability-matrix">
<h2>Method 1: Probability Matrix<a class="headerlink" href="#method-1-probability-matrix" title="Link to this heading">#</a></h2>
<p>If you’ve already calculated your genotype probabilities, you can input them directly using the <code class="docutils literal notranslate"><span class="pre">from_probs</span></code> method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">scistree</span><span class="o">.</span><span class="n">probability</span><span class="o">.</span><span class="n">from_probs</span><span class="p">(</span><span class="n">probs</span><span class="p">,</span> <span class="n">cell_names</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">site_names</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">margin</span><span class="o">=</span><span class="mf">1e-5</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">probs</span></code></strong>: A NumPy array containing the genotype probabilities.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">cell_names</span></code> (Optional)</strong>: A list of names for your cells. If you don’t provide this, cells will be automatically named <span class="math notranslate nohighlight">\(c_1\)</span>, <span class="math notranslate nohighlight">\(c_2\)</span>, up to <span class="math notranslate nohighlight">\(c_N\)</span>.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">site_names</span></code> (Optional)</strong>: A list of names for your genomic sites. If not provided, sites will be named <span class="math notranslate nohighlight">\(s_1\)</span>, <span class="math notranslate nohighlight">\(s_2\)</span>, up to <span class="math notranslate nohighlight">\(s_N\)</span>.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">margin</span></code></strong>: This parameter helps prevent probabilities from becoming either extremely large or extremely small, ensuring numerical stability.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method is ideal when you have precise genotype probabilities readily available, perhaps from another analysis pipeline or a simulation.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scistree2</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">s2</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span> 

<span class="n">probs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mf">0.08</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">],</span>
                 <span class="p">[</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.02</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">],</span>
                 <span class="p">[</span><span class="mf">0.02</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.02</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">],</span>
                 <span class="p">[</span><span class="mf">0.9</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.02</span><span class="p">],</span>
                 <span class="p">[</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">],</span>
                 <span class="p">[</span><span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.02</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">]])</span> 
<span class="n">cell_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cell1&#39;</span><span class="p">,</span> <span class="s1">&#39;cell2&#39;</span><span class="p">,</span> <span class="s1">&#39;cell3&#39;</span><span class="p">,</span> <span class="s1">&#39;cell4&#39;</span><span class="p">,</span> <span class="s1">&#39;cell5&#39;</span><span class="p">]</span>
<span class="n">site_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;snp1&#39;</span><span class="p">,</span> <span class="s1">&#39;snp2&#39;</span><span class="p">,</span> <span class="s1">&#39;snp3&#39;</span><span class="p">,</span> <span class="s1">&#39;snp4&#39;</span><span class="p">,</span> <span class="s1">&#39;snp5&#39;</span><span class="p">,</span> <span class="s1">&#39;snp6&#39;</span><span class="p">]</span>
<span class="n">gp</span> <span class="o">=</span> <span class="n">s2</span><span class="o">.</span><span class="n">probability</span><span class="o">.</span><span class="n">from_probs</span><span class="p">(</span><span class="n">probs</span><span class="p">,</span> <span class="n">cell_names</span><span class="p">,</span> <span class="n">site_names</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gp</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>      cell1  cell2  cell3  cell4  cell5
snp1   0.01   0.60   0.08   0.80   0.70
snp2   0.80   0.02   0.70   0.01   0.30
snp3   0.02   0.80   0.02   0.80   0.90
snp4   0.90   0.90   0.80   0.80   0.02
snp5   0.01   0.80   0.01   0.80   0.90
snp6   0.05   0.02   0.70   0.05   0.90
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-2-from-csv-file">
<h2>Method 2: From CSV File<a class="headerlink" href="#method-2-from-csv-file" title="Link to this heading">#</a></h2>
<p>If you’ve already had a csv file containing probabilites or reads information, you can input it directly using the <code class="docutils literal notranslate"><span class="pre">from_csv</span></code> method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">scistree</span><span class="o">.</span><span class="n">probability</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span><span class="n">csv_filepath</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="s2">&quot;probability&quot;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">csv_path</span></code></strong>: The path to your CSV file.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">source</span></code></strong>: Be either “probability” or “read”.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">kwargs</span></code></strong>: Other parameters, such as <code class="docutils literal notranslate"><span class="pre">ado</span></code>, <code class="docutils literal notranslate"><span class="pre">seqerr</span></code>, etc.</p></li>
</ul>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p></p></th>
<th class="head text-right"><p>cell1</p></th>
<th class="head text-right"><p>cell2</p></th>
<th class="head text-right"><p>cell3</p></th>
<th class="head text-right"><p>cell4</p></th>
<th class="head text-right"><p>cell5</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>snp1</p></td>
<td class="text-right"><p>0.01</p></td>
<td class="text-right"><p>0.6</p></td>
<td class="text-right"><p>0.08</p></td>
<td class="text-right"><p>0.8</p></td>
<td class="text-right"><p>0.7</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>snp2</p></td>
<td class="text-right"><p>0.8</p></td>
<td class="text-right"><p>0.02</p></td>
<td class="text-right"><p>0.7</p></td>
<td class="text-right"><p>0.01</p></td>
<td class="text-right"><p>0.3</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>snp3</p></td>
<td class="text-right"><p>0.02</p></td>
<td class="text-right"><p>0.8</p></td>
<td class="text-right"><p>0.02</p></td>
<td class="text-right"><p>0.8</p></td>
<td class="text-right"><p>0.9</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>snp4</p></td>
<td class="text-right"><p>0.9</p></td>
<td class="text-right"><p>0.9</p></td>
<td class="text-right"><p>0.8</p></td>
<td class="text-right"><p>0.8</p></td>
<td class="text-right"><p>0.02</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>snp5</p></td>
<td class="text-right"><p>0.01</p></td>
<td class="text-right"><p>0.8</p></td>
<td class="text-right"><p>0.01</p></td>
<td class="text-right"><p>0.8</p></td>
<td class="text-right"><p>0.9</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>snp6</p></td>
<td class="text-right"><p>0.05</p></td>
<td class="text-right"><p>0.02</p></td>
<td class="text-right"><p>0.7</p></td>
<td class="text-right"><p>0.05</p></td>
<td class="text-right"><p>0.9</p></td>
</tr>
</tbody>
</table>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gp</span> <span class="o">=</span> <span class="n">s2</span><span class="o">.</span><span class="n">probability</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span><span class="s1">&#39;./data/toy_probs.csv&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gp</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>      cell1  cell2  cell3  cell4  cell5
snp1   0.01   0.60   0.08   0.80   0.70
snp2   0.80   0.02   0.70   0.01   0.30
snp3   0.02   0.80   0.02   0.80   0.90
snp4   0.90   0.90   0.80   0.80   0.02
snp5   0.01   0.80   0.01   0.80   0.90
snp6   0.05   0.02   0.70   0.05   0.90
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-3-from-read-counts">
<h2>Method 3: From Read Counts<a class="headerlink" href="#method-3-from-read-counts" title="Link to this heading">#</a></h2>
<p>For raw read count data, ScisTree2 can automatically convert these into genotype probabilities. This method includes crucial parameters to account for common experimental noise in sequencing data.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">scistree</span><span class="o">.</span><span class="n">probability</span><span class="o">.</span><span class="n">from_reads</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">ado</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">seqerr</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">posterior</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">af</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cell_names</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">site_names</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">reads</span></code></strong>: A three-dimensional NumPy array containing the read counts for each cell at each site.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">ado</span></code></strong>: The Allele Dropout (ADO) rate, which accounts for the probability of an allele not being detected in a cell.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">seqerr</span></code></strong>: The sequencing error rate, representing the probability of a base being incorrectly called during sequencing.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">posterior</span></code></strong>: If set to <code class="docutils literal notranslate"><span class="pre">True</span></code>, ScisTree2 will use <code class="docutils literal notranslate"><span class="pre">af</span></code> to calculate the posterior genotype probability. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, <code class="docutils literal notranslate"><span class="pre">af</span></code> defaults to <code class="docutils literal notranslate"><span class="pre">0.5</span></code>.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">af</span></code></strong>: The allele frequency.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">cell_names</span></code> (Optional)</strong>: A list of names for your cells. If you don’t provide this, cells will be automatically named <code class="docutils literal notranslate"><span class="pre">c1</span></code>, <code class="docutils literal notranslate"><span class="pre">c2</span></code>, up to <code class="docutils literal notranslate"><span class="pre">cN</span></code>.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">site_names</span></code> (Optional)</strong>: A list of names for your genomic sites. If not provided, sites will be named <code class="docutils literal notranslate"><span class="pre">s1</span></code>, <code class="docutils literal notranslate"><span class="pre">s2</span></code>, up to <code class="docutils literal notranslate"><span class="pre">sN</span></code>.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is your go-to method when you have sequencing read depth information and need to infer genotype probabilities while accounting for common single-cell sequencing artifacts.</p>
</div>
<p>In this dataset, the input format remains a matrix of shape <code class="docutils literal notranslate"><span class="pre">(num_sites,</span> <span class="pre">num_cells)</span></code>.<br />
However, instead of using <strong>precomputed genotype probabilities</strong>, <strong>each entry is a tuple</strong> representing raw sequencing read counts.</p>
<p>*if the read counts are unknown, just use (0, 0).</p>
<p>Each tuple has the form: <code class="docutils literal notranslate"><span class="pre">(ref_count,</span> <span class="pre">alt_count)</span></code>, where:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">ref_count</span></code> is the number of reads supporting the <strong>reference (wild type)</strong> allele</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alt_count</span></code> is the number of reads supporting the <strong>mutation (alternative)</strong> allele</p></li>
</ul>
<p>This format is suitable when you start from <strong>raw read counts</strong> rather than inferred genotype probabilities, enabling ScisTree2 to perform <strong>probabilistic genotype modeling</strong> internally before tree inference.</p>
<p>This dataset, located in the <code class="docutils literal notranslate"><span class="pre">data</span></code> folder, contains <strong>50 cells</strong> and <strong>100 SNPs</strong>.<br />
It was generated using <a class="reference external" href="https://github.com/dapogon/cellcoal">CellCoal</a> with the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cellcoal-1.2.0<span class="w"> </span>-n5<span class="w"> </span>-s10<span class="w"> </span>-l100<span class="w"> </span>-e1000<span class="w"> </span>-b1<span class="w"> </span>-j30<span class="w"> </span>-p0<span class="w"> </span>-D0.5<span class="w"> </span>-B0<span class="w"> </span><span class="m">0</span>.01<span class="w"> </span>-C5<span class="w"> </span>-E0<span class="w"> </span>-otoys<span class="w"> </span>-y3<span class="w"> </span>-v<span class="w"> </span>-1<span class="w"> </span>-2<span class="w"> </span>-6<span class="w"> </span>-7<span class="w"> </span>-9<span class="w"> </span>-Y
</pre></div>
</div>
<p>Preview:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p>Cell 1</p></th>
<th class="head"><p>Cell 2</p></th>
<th class="head"><p>Cell 3</p></th>
<th class="head"><p>Cell 4</p></th>
<th class="head"><p>Cell 5</p></th>
<th class="head"><p>Cell 6</p></th>
<th class="head"><p>Cell 7</p></th>
<th class="head"><p>Cell 8</p></th>
<th class="head"><p>Cell 9</p></th>
<th class="head"><p>Cell 10</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>SNP 1</p></td>
<td><p>(4,1)</p></td>
<td><p>(4,0)</p></td>
<td><p>(5,1)</p></td>
<td><p>(4,0)</p></td>
<td><p>(1,0)</p></td>
<td><p>(4,0)</p></td>
<td><p>(6,0)</p></td>
<td><p>(6,0)</p></td>
<td><p>(5,0)</p></td>
<td><p>(5,0)</p></td>
</tr>
<tr class="row-odd"><td><p>SNP 2</p></td>
<td><p>(4,0)</p></td>
<td><p>(3,0)</p></td>
<td><p>(7,1)</p></td>
<td><p>(1,1)</p></td>
<td><p>(3,0)</p></td>
<td><p>(2,0)</p></td>
<td><p>(10,0)</p></td>
<td><p>(11,0)</p></td>
<td><p>(1,0)</p></td>
<td><p>(9,0)</p></td>
</tr>
<tr class="row-even"><td><p>SNP 3</p></td>
<td><p>(4,5)</p></td>
<td><p>(11,0)</p></td>
<td><p>(4,0)</p></td>
<td><p>(3,0)</p></td>
<td><p>(4,0)</p></td>
<td><p>(3,0)</p></td>
<td><p>(7,0)</p></td>
<td><p>(0,0)</p></td>
<td><p>(2,0)</p></td>
<td><p>(4,0)</p></td>
</tr>
<tr class="row-odd"><td><p>SNP 4</p></td>
<td><p>(8,0)</p></td>
<td><p>(2,0)</p></td>
<td><p>(6,0)</p></td>
<td><p>(1,0)</p></td>
<td><p>(2,2)</p></td>
<td><p>(3,5)</p></td>
<td><p>(5,3)</p></td>
<td><p>(1,3)</p></td>
<td><p>(2,4)</p></td>
<td><p>(8,0)</p></td>
</tr>
<tr class="row-even"><td><p>SNP 5</p></td>
<td><p>(5,0)</p></td>
<td><p>(9,0)</p></td>
<td><p>(4,0)</p></td>
<td><p>(3,0)</p></td>
<td><p>(4,0)</p></td>
<td><p>(3,0)</p></td>
<td><p>(7,0)</p></td>
<td><p>(4,1)</p></td>
<td><p>(4,0)</p></td>
<td><p>(5,0)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only the first 10 cells and 5 SNPs are shown here for illustration. The full dataset contains 100 SNPs × 50 cells. Use (0, 0) if there is any missing.</p>
</div>
<p>We calculate the posterior genotype probability with the following settings:</p>
<ul class="simple">
<li><p>Allelic dropout rate: <code class="docutils literal notranslate"><span class="pre">ado=0.2</span></code></p></li>
<li><p>Sequencing error rate: <code class="docutils literal notranslate"><span class="pre">seqerr=0.01</span></code></p></li>
</ul>
<p>When <code class="docutils literal notranslate"><span class="pre">posterior=True</span></code> (default), the posterior probability <span class="math notranslate nohighlight">\(p(G \mid D)\)</span> is calculated given the allele frequency <code class="docutils literal notranslate"><span class="pre">af</span></code>.<br />
The parameter <code class="docutils literal notranslate"><span class="pre">af</span></code> should be a <code class="docutils literal notranslate"><span class="pre">numpy.ndarray</span></code> with shape (num_sites, 1).</p>
<p>If <code class="docutils literal notranslate"><span class="pre">af</span></code> is not provided or set to <code class="docutils literal notranslate"><span class="pre">None</span></code>, it will be automatically estimated from the current samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># one can generate the input directly from a reads matrix.</span>
<span class="n">reads</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;data/toy_raw_reads.npy&#39;</span><span class="p">,</span> <span class="n">allow_pickle</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># load simulated reads</span>
<span class="n">cell_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;cell</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">50</span><span class="p">)]</span>
<span class="n">site_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;snp</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">)]</span>
<span class="n">gp</span> <span class="o">=</span> <span class="n">s2</span><span class="o">.</span><span class="n">probability</span><span class="o">.</span><span class="n">from_reads</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">ado</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">seqerr</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">posterior</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">af</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cell_names</span><span class="o">=</span><span class="n">cell_names</span><span class="p">,</span> <span class="n">site_names</span><span class="o">=</span><span class="n">site_names</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gp</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>       cell0  cell1  cell2  cell3  cell4  cell5  cell6  cell7  cell8  cell9  \
snp0    0.97   0.93   0.97   0.97   0.95   0.93   0.50   0.97   0.96   0.74   
snp1    0.98   0.71   0.82   0.98   0.82   0.93   0.96   0.98   0.98   0.97   
snp2    0.44   0.96   0.28   0.95   0.95   0.97   0.96   0.96   0.96   0.95   
snp3    0.74   0.88   0.88   0.01   0.00   0.00   0.34   0.90   0.01   0.88   
snp4    0.01   0.00   0.00   0.89   0.46   0.91   0.04   0.00   0.06   0.00   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
snp95   0.88   0.93   0.92   0.83   0.97   0.92   0.58   0.83   0.96   0.97   
snp96   0.87   0.94   0.00   0.94   0.97   0.87   0.96   0.97   0.55   0.92   
snp97   0.00   0.50   0.00   0.85   0.30   0.89   0.00   0.10   0.21   0.00   
snp98   0.99   0.99   0.96   0.98   0.98   0.99   0.98   0.98   0.99   0.98   
snp99   0.98   0.62   0.62   0.97   0.94   0.98   0.90   0.94   0.98   0.96   

       ...  cell40  cell41  cell42  cell43  cell44  cell45  cell46  cell47  \
snp0   ...    0.93    0.97    0.44    0.96    0.28    0.96    0.96    0.44   
snp1   ...    0.98    0.98    0.97    0.98    0.71    0.50    0.98    0.98   
snp2   ...    0.50    0.50    0.28    0.97    0.95    0.28    0.96    0.50   
snp3   ...    0.06    0.12    0.00    0.00    0.00    0.82    0.88    0.65   
snp4   ...    0.86    0.71    0.92    0.89    0.89    0.06    0.92    0.86   
...    ...     ...     ...     ...     ...     ...     ...     ...     ...   
snp95  ...    0.50    0.97    0.97    0.18    0.97    0.50    0.96    0.95   
snp96  ...    0.96    0.94    0.14    0.87    0.97    0.97    0.94    0.01   
snp97  ...    0.79    0.50    0.90    0.88    0.70    0.00    0.92    0.60   
snp98  ...    0.36    0.99    0.99    0.99    0.97    0.99    0.99    0.98   
snp99  ...    0.90    0.06    0.96    0.97    0.85    0.97    0.90    0.06   

       cell48  cell49  
snp0     0.98    0.96  
snp1     0.98    0.98  
snp2     0.98    0.93  
snp3     0.74    0.00  
snp4     0.06    0.92  
...       ...     ...  
snp95    0.97    0.97  
snp96    0.09    0.96  
snp97    0.00    0.70  
snp98    0.98    0.98  
snp99    0.94    0.94  

[100 rows x 50 columns]
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-4-from-vcf-file">
<h2>Method 4: From VCF File<a class="headerlink" href="#method-4-from-vcf-file" title="Link to this heading">#</a></h2>
<p>ScisTree2 also supports direct input from Variant Call Format (VCF) files, a common format for storing genetic variation data.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">scistree</span><span class="o">.</span><span class="n">probability</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="n">vcf_path</span><span class="p">,</span> <span class="n">ado</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">seqerr</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">posterior</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">af</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s1">&#39;AD&#39;</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">vcf_path</span></code></strong>: The path to your VCF file.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">ado</span></code></strong>: The Allele Dropout (ADO) rate, which accounts for the probability of an allele not being detected in a cell.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">seqerr</span></code></strong>: The sequencing error rate, representing the probability of a base being incorrectly called during sequencing.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">posterior</span></code></strong>: If set to True, ScisTree2 will use <code class="docutils literal notranslate"><span class="pre">af</span></code> to calculate the posterior genotype probability. If False, <code class="docutils literal notranslate"><span class="pre">af</span></code> defaults to <span class="math notranslate nohighlight">\(0.5\)</span>.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">af</span></code></strong>: The allele frequency.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">key</span></code></strong>: This is the specific tag within the VCF file that stores the read depth information.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vcf_path</span> <span class="o">=</span> <span class="s1">&#39;./data/example.vcf&#39;</span>
<span class="n">gp</span> <span class="o">=</span> <span class="n">s2</span><span class="o">.</span><span class="n">probability</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="n">vcf_path</span><span class="p">,</span> <span class="n">ado</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">seqerr</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">posterior</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">af</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gp</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Info: Skipping line because AD not found in FORMAT: CHR[chr1], POS[250]
          CELL_1  CELL_2  CELL_3  CELL_4  CELL_5
chr1:100    0.00    0.88    0.00    0.00    0.88
chr1:150    0.95    0.00    0.50    0.95    0.00
chr1:200    0.00    0.00    0.77    0.00    0.00
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./tutorials"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="usage.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Basic usage</p>
      </div>
    </a>
    <a class="right-next"
       href="branch_length.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Map mutations on branches</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method-1-probability-matrix">Method 1: Probability Matrix</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method-2-from-csv-file">Method 2: From CSV File</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method-3-from-read-counts">Method 3: From Read Counts</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method-4-from-vcf-file">Method 4: From VCF File</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Haotian Zhang, Yufeng Wu
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025 LabWu @ University of Connecticut.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>